src/Streamly/Internal/Data/Stream/StreamD/Type.hs

```diff
@@ -289,6 +290,30 @@ foldOnce (Fold fstep begin done) (Stream step state) = do
             Skip s -> go SPEC fs s
             Stop -> done fs
 
+{-# INLINE_NORMAL foldAndReturn #-}
+foldAndReturn :: Monad m => Fold m a b -> Stream m a -> m (b, Stream m a)
```

Comment:
If this is as efficient as `fold` (which I believe it will be), you can implement `fold` interms of `foldAndReturn`

Reply:
We should also change the name foldOnce to fold.